maestro/lifecycle.py

```diff
@@ -108,12 +111,31 @@ def _create_env(self):
         return dict((str(k), str(v)) for k, v in env.items())
 
     def _test(self, container=None):
-        return subprocess.call(self.command, env=self._create_env()) == 0
+        env = self._create_env()
+        if self.envfrom == "env":
+            return subprocess.call(self.command, env=env) == 0
+        elif self.envfrom == 'stdin':
+            file_name = "/tmp/stdin_{}_{}".format(
+                threading.current_thread().ident, time.time(),
+            )
+            try:
+                with open(file_name, "w") as f:
+                    json.dump(env, f)
+                command = self.command + ["-f", file_name]
+                return subprocess.call(command) == 0
```

Comment:
What you're doing here is not passing data through stdin. You can simply do this:

```python
import shlex

def _test(self, container=None):
    env = self._create_env()
    if self.envfrom == 'env':
        return subprocess...
    elif self.envfrom == 'stdin':
        p = subprocess.Popen(shlex.split(self.command), stdin=subprocess.PIPE)
        p.communicate(json.dumps(env).encode('utf-8'))
        return p.wait() == 0
    else:
        raise ValueError(self.envfrom)
```

This avoids the management of the temporary file (which you should have done with the `tempfile` module anyway), and makes it a cleaner interface for the scripts/commands that are executed -- not all of them may be able to accept a `-f <file>` extra argument.

You're also using JSON as the encoding for the data, which is quite different, although I like that it removes concerns around escaping of complex values, so let's keep that.

Reply:
Thanks a lot Max for the suggestion. It is surely much cleaner. I have made the change accordingly.

Reply:
Also use subprocess.Popen(stdin=PIPE) and subprocess.Popen.communicate(input=None) instead of subprocess.call() to make sure health checking script handles envfrom='env' correctly